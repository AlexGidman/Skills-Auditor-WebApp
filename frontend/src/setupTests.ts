// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
