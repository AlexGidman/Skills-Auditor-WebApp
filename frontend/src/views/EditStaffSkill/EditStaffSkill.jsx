import React, { useState, useEffect } from "react";
import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { TextAreaInput } from "../../components/TextAreaInput/TextAreaInput";
import { Select } from "../../components/Select/Select";
import { Button } from "../../components/Button/Button";

import { useAPI, getSelectOptionsFromArray, formatDate } from "../../utility/helper";
import { getStaffSkill, updateStaffSkill, getAllDirectReports } from "../../utility/apiRequests";
import styles from "./EditStaffSkill.module.css";

import { useNavigate, useParams, useOutletContext } from "react-router-dom";
import { SKILL_LEVELS } from "../../utility/types";

export const EditStaffSkill = () => {
    const { staffSkillId } = useParams();
    const [currentUser, setShowToast] = useOutletContext();
    const navigate = useNavigate();

    const { data, loading, error } = useAPI(getStaffSkill, [staffSkillId]);

    const {
        data: directReportsData,
        loading: directReportsLoading,
        error: directReportsError,
    } = useAPI(getAllDirectReports, [currentUser.id]);

    useEffect(() => {
        if (data && currentUser.id !== data.user_id) {
            if (
                !!directReportsError ||
                (!!directReportsData &&
                    !directReportsData.filter((directReport) => {
                        return directReport.report.id.toString() === data.user_id;
                    }))
            ) {
                setShowToast({
                    error: "You are not authorised to edit this user's skills",
                });
                navigate("/");
            }
        }
    }, [currentUser.id, data, directReportsData, directReportsError, navigate, setShowToast]);

    if (loading || directReportsLoading) return <LoadingPlaceholder />;
    if (error) return <ErrorMessage error={error} />;
    return (
        <>
            <header>
                <h1>Edit Staff Skill</h1>
            </header>
            <section>
                <Form data={data} />
            </section>
        </>
    );
};

const Form = ({ data }) => {
    const navigate = useNavigate();
    const [, setShowToast] = useOutletContext();
    const [skillLevel, setSkillLevel] = useState(data.skill_level);
    const [skillNotes, setSkillNotes] = useState(data.notes);

    const [formChanged, setFormChanged] = useState(false);
    const [navigateBack, setNavigateBack] = useState(false);

    const {
        error,
        loading,
        callback: updateStaffSkillInfo,
    } = useAPI(updateStaffSkill, [data.id, skillLevel, skillNotes], true);

    useEffect(() => {
        setFormChanged(skillLevel !== data.skill_level || skillNotes !== data.notes);
    }, [data, skillLevel, skillNotes]);

    useEffect(() => {
        if (navigateBack && !loading) {
            error
                ? setShowToast({ error: "Failed to update staff skill" })
                : setShowToast({ success: "Successfully updated staff skill" });
            navigate(-1);
        }
    }, [navigateBack, loading, error, setShowToast, navigate]);

    return (
        <form className={styles.form}>
            <span
                role="status"
                className={styles.item}
                aria-label={
                    data.expiry_date
                        ? `${data.skillID.name} (Expires ${formatDate(data.expiry_date)})`
                        : data.skillID.name
                }
            >
                {data.skillID.name}
                <em>{data.expiry_date ? ` (Expires ${formatDate(data.expiry_date)})` : ""}</em>
            </span>
            <Select
                className={styles.item}
                labelText="Select a level"
                options={getSelectOptionsFromArray([1, 2, 3, 4, 5], SKILL_LEVELS)}
                value={skillLevel}
                onChange={(e) => {
                    setSkillLevel(e.target.value);
                }}
            />
            <TextAreaInput
                labelText="Notes"
                className={styles.item}
                value={skillNotes}
                onChange={(e) => {
                    setSkillNotes(e.target.value);
                }}
            />
            <div className={styles.buttonWrapper}>
                <Button
                    className={styles.button}
                    loading={loading}
                    disabled={!formChanged}
                    onClick={(e) => {
                        e.preventDefault();
                        updateStaffSkillInfo();
                        setNavigateBack(true);
                    }}
                >
                    Save changes
                </Button>
                <Button
                    className={styles.button}
                    onClick={(e) => {
                        e.preventDefault();
                        navigate(-1);
                    }}
                >
                    Cancel
                </Button>
            </div>
        </form>
    );
};
