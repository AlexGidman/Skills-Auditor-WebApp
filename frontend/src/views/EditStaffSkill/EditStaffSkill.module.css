.button {
    min-width: 16rem;
    flex: 44%;
}

.buttonWrapper {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-between;
    gap: var(--s2);
}

.form {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    margin: 0 20%;
}

.header {
    margin-top: 10%;
}

.item {
    margin-bottom: var(--s1);
    min-width: 14rem;
}

@media (max-width: 900px) {
    .form {
        margin: 0;
    }
}
