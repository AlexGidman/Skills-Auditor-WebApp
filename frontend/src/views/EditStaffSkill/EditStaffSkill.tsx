import React, { useState, useEffect } from "react";
import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { TextAreaInput } from "../../components/TextAreaInput/TextAreaInput";
import { Select } from "../../components/Select/Select";
import { Button } from "../../components/Button/Button";

import { useAPI, getSelectOptionsFromArray, formatDate } from "../../utility/helper";
import { getStaffSkill, updateStaffSkill, getAllDirectReports } from "../../utility/apiRequests";
import styles from "./EditStaffSkill.module.css";

import { useNavigate, useParams, useOutletContext } from "react-router-dom";
import { DirectReport, SKILL_LEVELS, StaffSkill } from "../../utility/types";
import { AppOutletContext } from "../AppWrapper/AppWrapper";

export const EditStaffSkill = () => {
    const { staffskillId } = useParams();
    const { currentUser, setShowToast } = useOutletContext<AppOutletContext>();

    const navigate = useNavigate();

    const { data, loading, error } = useAPI<StaffSkill>(getStaffSkill, [staffskillId]);

    const {
        data: directReportsData,
        loading: directReportsLoading,
        error: directReportsError,
    } = useAPI<DirectReport[]>(getAllDirectReports, [currentUser.id]);

    useEffect(() => {
        if (data && currentUser.id !== data.userId) {
            if (
                !!directReportsError ||
                (!!directReportsData &&
                    !directReportsData.filter((directReport) => {
                        return directReport?.report?.id?.toString() === data.userId;
                    }))
            ) {
                setShowToast({
                    error: "You are not authorised to edit this user's skills",
                });
                navigate("/");
            }
        }
    }, [currentUser.id, data, directReportsData, directReportsError, navigate, setShowToast]);

    if (loading || directReportsLoading) return <LoadingPlaceholder />;
    if (error) return <ErrorMessage error={error} />;

    return (
        <>
            <header>
                <h1>Edit Staff Skill</h1>
            </header>
            <section>{data && <Form data={data} />}</section>
        </>
    );
};

interface FormProps {
    data: StaffSkill;
}

const Form = ({ data }: FormProps) => {
    const navigate = useNavigate();
    const { setShowToast } = useOutletContext<AppOutletContext>();

    const [skillLevel, setSkillLevel] = useState(data.skillLevel);
    const [skillNotes, setSkillNotes] = useState(data.notes);

    const [formChanged, setFormChanged] = useState(false);
    const [navigateBack, setNavigateBack] = useState(false);

    const {
        error,
        loading,
        callback: updateStaffSkillInfo,
    } = useAPI(updateStaffSkill, [data.id, skillLevel, skillNotes], true);

    useEffect(() => {
        setFormChanged(skillLevel !== data.skillLevel || skillNotes !== data.notes);
    }, [data, skillLevel, skillNotes]);

    useEffect(() => {
        if (navigateBack && !loading) {
            error
                ? setShowToast({ error: "Failed to update staff skill" })
                : setShowToast({ success: "Successfully updated staff skill" });
            navigate(-1);
        }
    }, [navigateBack, loading, error, setShowToast, navigate]);

    return (
        <form className={styles.form}>
            <span
                role="status"
                className={styles.item}
                aria-label={
                    data.expiryDate
                        ? `${data.skill.name} (Expires ${formatDate(data.expiryDate)})`
                        : data.skill.name
                }
            >
                {data.skill.name}
                <em>{data.expiryDate ? ` (Expires ${formatDate(data.expiryDate)})` : ""}</em>
            </span>
            <Select
                className={styles.item}
                labelText="Select a level"
                options={getSelectOptionsFromArray(
                    ["1", "2", "3", "4", "5"],
                    Array.from(SKILL_LEVELS),
                )}
                value={skillLevel}
                onChange={(e) => {
                    setSkillLevel(e.target.value);
                }}
            />
            <TextAreaInput
                labelText="Notes"
                className={styles.item}
                value={skillNotes}
                onChange={(e) => {
                    setSkillNotes(e.target.value);
                }}
            />
            <div className={styles.buttonWrapper}>
                <Button
                    className={styles.button}
                    loading={loading}
                    disabled={!formChanged}
                    onClick={(e) => {
                        e.preventDefault();
                        updateStaffSkillInfo();
                        setNavigateBack(true);
                    }}
                >
                    Save changes
                </Button>
                <Button
                    className={styles.button}
                    onClick={(e) => {
                        e.preventDefault();
                        navigate(-1);
                    }}
                >
                    Cancel
                </Button>
            </div>
        </form>
    );
};
