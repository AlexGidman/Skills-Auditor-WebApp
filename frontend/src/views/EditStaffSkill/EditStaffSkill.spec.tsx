import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { mockAdminUser, mockApiRequests, mockApiResponse } from "../../setupTests";
import * as ReactDom from "react-router-dom";
import { EditStaffSkill } from "./EditStaffSkill";
import {
    BASIC,
    Category,
    EXPERT,
    HIGH,
    INTERMEDIATE,
    NONE,
    Skill,
    StaffSkill,
} from "../../utility/types";
import { wait } from "@testing-library/user-event/dist/utils";
import React from "react";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
        useNavigate: jest.fn(),
    };
});

const mockCategory: Category = { id: "1", name: "testname1" };

const mockSkill1: Skill = {
    id: "1",
    name: "Skill1",
    category: mockCategory,
    categoryId: mockCategory.id,
};

const mockStaffSkill1: StaffSkill = {
    id: "100",
    skillId: mockSkill1.id,
    skill: mockSkill1,
    skillLevel: "1",
    userId: mockAdminUser.id,
    notes: "test notes",
    expiryDate: "0000-00-00",
};

const mockStaffSkill2: StaffSkill = {
    id: "100",
    skillId: mockSkill1.id,
    skill: mockSkill1,
    skillLevel: "1",
    userId: mockAdminUser.id,
    notes: "test notes",
    expiryDate: "1973-01-01",
};

describe("EditStaffSkill", () => {
    beforeEach(() => {
        mockApiRequests.getAllDirectReports.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: [] }),
        );
    });
    it("renders correctly when data retrieval successful and skill has no expiry date", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });

        mockApiRequests.getStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: mockStaffSkill1 }),
        );
        const { container } = render(
            <ReactDom.MemoryRouter>
                <EditStaffSkill />
            </ReactDom.MemoryRouter>,
        );
        expect(
            await screen.findByRole("heading", { name: "Edit Staff Skill" }),
        ).toBeInTheDocument();

        expect(
            screen.getByRole("status", { name: mockStaffSkill1.skillId.name }),
        ).toBeInTheDocument();
        expect(screen.getByRole("combobox", { name: "Select a level" })).toBeInTheDocument();
        expect(screen.getByRole("combobox", { name: "Select a level" })).toHaveValue(
            mockStaffSkill1.skillLevel,
        );
        expect(screen.getByRole("option", { name: NONE })).toBeInTheDocument();
        expect(screen.getByRole("option", { name: BASIC })).toBeInTheDocument();
        expect(screen.getByRole("option", { name: INTERMEDIATE })).toBeInTheDocument();
        expect(screen.getByRole("option", { name: HIGH })).toBeInTheDocument();
        expect(screen.getByRole("option", { name: EXPERT })).toBeInTheDocument();

        expect(screen.getByRole("textbox", { name: "Notes:" })).toBeInTheDocument();
        expect(screen.getByRole("textbox", { name: "Notes:" })).toHaveValue(mockStaffSkill1.notes);

        expect(screen.getByRole("button", { name: "Save changes" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Save changes" })).toBeDisabled();
        expect(screen.getByRole("button", { name: "Cancel" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Cancel" })).not.toBeDisabled();
        expect(container).toMatchSnapshot();
    });

    it("renders correctly when data retrieval successful and skill ahs expiry date", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });

        mockApiRequests.getStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: mockStaffSkill2 }),
        );
        const { container } = render(
            <ReactDom.MemoryRouter>
                <EditStaffSkill />
            </ReactDom.MemoryRouter>,
        );
        expect(
            await screen.findByRole("heading", { name: "Edit Staff Skill" }),
        ).toBeInTheDocument();

        expect(
            screen.getByRole("status", {
                name: `${mockStaffSkill2.skill.name} (Expires 01/01/1973)`,
            }),
        ).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it("should have Save changes button that becomes enabled when form is valid and has changed", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });

        mockApiRequests.getStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: mockStaffSkill1 }),
        );
        render(
            <ReactDom.MemoryRouter>
                <EditStaffSkill />
            </ReactDom.MemoryRouter>,
        );
        expect(
            await screen.findByRole("heading", { name: "Edit Staff Skill" }),
        ).toBeInTheDocument();

        const inputChangeArray = [
            {
                change: () => userEvent.type(screen.getByRole("textbox", { name: "Notes:" }), "a"),
                undo: () =>
                    userEvent.type(screen.getByRole("textbox", { name: "Notes:" }), "{backspace}"),
            },

            {
                change: () =>
                    userEvent.selectOptions(
                        screen.getByRole("combobox", { name: "Select a level" }),
                        "2",
                    ),
                undo: () =>
                    userEvent.selectOptions(
                        screen.getByRole("combobox", { name: "Select a level" }),
                        mockStaffSkill1.skillLevel,
                    ),
            },
        ];
        for (let changeInputFunc of inputChangeArray) {
            expect(screen.getByRole("button", { name: "Save changes" })).toBeDisabled();
            changeInputFunc.change();
            expect(screen.getByRole("button", { name: "Save changes" })).not.toBeDisabled();
            changeInputFunc.undo();
            expect(screen.getByRole("button", { name: "Save changes" })).toBeDisabled();
        }
    });

    it("should pass skill params to updateStaffSkill API when Save changes button clicked", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });

        mockApiRequests.getStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: mockStaffSkill1 }),
        );

        mockApiRequests.updateStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: "Success" }),
        );
        render(
            <ReactDom.MemoryRouter>
                <EditStaffSkill />
            </ReactDom.MemoryRouter>,
        );
        expect(
            await screen.findByRole("heading", { name: "Edit Staff Skill" }),
        ).toBeInTheDocument();

        expect(mockApiRequests.updateStaffSkill).not.toHaveBeenCalled();
        userEvent.type(screen.getByRole("textbox", { name: "Notes:" }), "a");
        userEvent.click(screen.getByRole("button", { name: "Save changes" }));
        expect(mockApiRequests.updateStaffSkill).toHaveBeenCalled();
        expect(mockApiRequests.updateStaffSkill).toHaveBeenLastCalledWith(
            { headers: { authorization: "Bearer token" } },
            mockStaffSkill1.id,
            mockStaffSkill1.skillLevel,
            mockStaffSkill1.notes + "a",
        );
    });

    it("should redirect to previous page when save changes button clicked and update successful", async () => {
        const mockNavigate = jest.fn();
        jest.spyOn(ReactDom, "useNavigate").mockImplementation(() => mockNavigate);
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });

        mockApiRequests.getStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: mockStaffSkill1 }),
        );

        mockApiRequests.updateStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: "Success" }),
        );
        render(
            <ReactDom.MemoryRouter>
                <EditStaffSkill />
            </ReactDom.MemoryRouter>,
        );
        expect(
            await screen.findByRole("heading", { name: "Edit Staff Skill" }),
        ).toBeInTheDocument();

        expect(mockNavigate).not.toHaveBeenCalled();
        userEvent.type(screen.getByRole("textbox", { name: "Notes:" }), "a");
        userEvent.click(screen.getByRole("button", { name: "Save changes" }));
        await wait(100);
        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenLastCalledWith(-1);
    });

    it("should redirect to previous page when save changes button clicked and update fails", async () => {
        const mockNavigate = jest.fn();
        jest.spyOn(ReactDom, "useNavigate").mockImplementation(() => mockNavigate);
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });

        mockApiRequests.getStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: mockStaffSkill1 }),
        );

        mockApiRequests.updateStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: "Success" }),
        );
        render(
            <ReactDom.MemoryRouter>
                <EditStaffSkill />
            </ReactDom.MemoryRouter>,
        );
        expect(
            await screen.findByRole("heading", { name: "Edit Staff Skill" }),
        ).toBeInTheDocument();

        expect(mockNavigate).not.toHaveBeenCalled();
        userEvent.type(screen.getByRole("textbox", { name: "Notes:" }), "a");
        userEvent.click(screen.getByRole("button", { name: "Save changes" }));
        await wait(100);
        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenLastCalledWith(-1);
    });

    it("redirects to previous page on click of Cancel button", async () => {
        const mockNavigate = jest.fn();
        jest.spyOn(ReactDom, "useNavigate").mockImplementation(() => mockNavigate);

        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });

        mockApiRequests.getStaffSkill.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: mockStaffSkill1 }),
        );
        render(
            <ReactDom.MemoryRouter>
                <EditStaffSkill />
            </ReactDom.MemoryRouter>,
        );

        expect(await screen.findByRole("button", { name: "Cancel" })).toBeInTheDocument();
        expect(mockNavigate).not.toHaveBeenCalled();

        userEvent.click(screen.getByRole("button", { name: "Cancel" }));
        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenLastCalledWith(-1);
    });
});
