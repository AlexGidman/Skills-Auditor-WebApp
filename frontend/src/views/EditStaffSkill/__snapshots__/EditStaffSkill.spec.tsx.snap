// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditStaffSkill renders correctly when data retrieval successful and skill ahs expiry date 1`] = `
<div>
  <header>
    <h1>
      Edit Staff Skill
    </h1>
  </header>
  <section>
    <form
      class="form"
    >
      <span
        aria-label="Skill1 (Expires 01/01/1973)"
        class="item"
        role="status"
      >
        Skill1
        <em>
           (Expires 01/01/1973)
        </em>
      </span>
      <div
        class="component item"
      >
        <label
          class="label"
          for="Select-a level-undefined"
        >
          Select a level
          :
        </label>
        <select
          aria-label="Select a level"
          class="select"
          name="Select a level"
        >
          <option
            value="1"
          >
            None
          </option>
          <option
            value="2"
          >
            Basic
          </option>
          <option
            value="3"
          >
            Intermediate
          </option>
          <option
            value="4"
          >
            High
          </option>
          <option
            value="5"
          >
            Expert
          </option>
        </select>
      </div>
      <div
        class="component item"
      >
        <label
          class="label"
          for="Notes-undefined"
        >
          Notes
          :
        </label>
        <textarea
          class="input"
          id="Notes-undefined"
          maxlength="100"
          minlength="0"
          rows="4"
        >
          test notes
        </textarea>
      </div>
      <div
        class="buttonWrapper"
      >
        <button
          class="component button"
          disabled=""
        >
          Save changes
        </button>
        <button
          class="component button"
        >
          Cancel
        </button>
      </div>
    </form>
  </section>
</div>
`;

exports[`EditStaffSkill renders correctly when data retrieval successful and skill has no expiry date 1`] = `
<div>
  <header>
    <h1>
      Edit Staff Skill
    </h1>
  </header>
  <section>
    <form
      class="form"
    >
      <span
        aria-label="Skill1 (Expires N/A)"
        class="item"
        role="status"
      >
        Skill1
        <em>
           (Expires N/A)
        </em>
      </span>
      <div
        class="component item"
      >
        <label
          class="label"
          for="Select-a level-undefined"
        >
          Select a level
          :
        </label>
        <select
          aria-label="Select a level"
          class="select"
          name="Select a level"
        >
          <option
            value="1"
          >
            None
          </option>
          <option
            value="2"
          >
            Basic
          </option>
          <option
            value="3"
          >
            Intermediate
          </option>
          <option
            value="4"
          >
            High
          </option>
          <option
            value="5"
          >
            Expert
          </option>
        </select>
      </div>
      <div
        class="component item"
      >
        <label
          class="label"
          for="Notes-undefined"
        >
          Notes
          :
        </label>
        <textarea
          class="input"
          id="Notes-undefined"
          maxlength="100"
          minlength="0"
          rows="4"
        >
          test notes
        </textarea>
      </div>
      <div
        class="buttonWrapper"
      >
        <button
          class="component button"
          disabled=""
        >
          Save changes
        </button>
        <button
          class="component button"
        >
          Cancel
        </button>
      </div>
    </form>
  </section>
</div>
`;
